Source/Csla/Core/GraphMerger.cs

```diff
@@ -202,5 +205,178 @@ private void MergeGraph(IEditableCollection target, IEditableCollection source)
+          if (sourceValue is IEditableBusinessObject sourceChild)
+          {
+            var targetFieldExists = true;
+            if (targetFieldManager != null && (item.RelationshipType & RelationshipTypes.LazyLoad) == RelationshipTypes.LazyLoad)
+              targetFieldExists = targetFieldManager.FieldExists(item);
+            if (targetFieldExists && ReadProperty(target, item) is IEditableBusinessObject targetChild)
+            {
+              await MergeGraphAsync(targetChild, sourceChild);
+            }
+            else
+            {
+              if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceChild);
+              else
+                LoadProperty(target, item, sourceChild);
+            }
+          }
+          else
+          {
+            if (sourceValue is IEditableCollection sourceList)
+            {
+              var targetFieldExists = true;
+              if (targetFieldManager != null && (item.RelationshipType & RelationshipTypes.LazyLoad) == RelationshipTypes.LazyLoad)
+                targetFieldExists = targetFieldManager.FieldExists(item);
+              if (targetFieldExists && ReadProperty(target, item) is IEditableCollection targetList)
+              {
+                await MergeGraphAsync(targetList, sourceList);
+              }
+              else
+              {
+                if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                  Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceList);
+                else
+                 LoadProperty(target, item, sourceList);
+              }
+            }
+            else
+            {
+              if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceValue);
+              else if (sourceValue != null || (item.RelationshipType & RelationshipTypes.LazyLoad) != RelationshipTypes.LazyLoad)
+                LoadProperty(target, item, sourceValue);
+            }
+          }
+        }
+        if (source.IsNew)
+        {
+          MarkNew(target);
+        }
+        else if (!source.IsDirty)
+        {
+          MarkOld(target);
+        }
+        else
+        {
+          CopyField(source, target, "_isDirty");
+          CopyField(source, target, "_isNew");
+          CopyField(source, target, "_isDeleted");
+        }
+        await CheckRulesAsync(target);
+      }
+    }
+    /// <summary>
+    /// Merges state from source graph into target graph.
+    /// </summary>
+    /// <param name="target">Target of merge.</param>
+    /// <param name="source">Source for merge.</param>
+    private async Task MergeGraphAsync(IEditableCollection target, IEditableCollection source)
```

Comment:
What about rewriting this method as
```csharp
var cacheKey = GetCacheKey(target);

var methodReference = _methodCache.GetOrAdd(cacheKey, ((Type ListType, Type ChildType, bool IsExtendedBindingList) key) =>
{
  var methodName = key.IsExtendedBindingList ? "MergeBusinessBindingListGraphAsync" : "MergeBusinessListGraphAsync";
  return GetType().GetMethod(methodName).MakeGenericMethod(key.ListType, key.ChildType);
});

var task = (Task)methodReference.Invoke(this, [target, source]);
await task;

static (Type ListType, Type ChildType, bool IsExtendedBindingList) GetCacheKey(IEditableCollection target)
{
  var listType = target.GetType();
  var childType = Utilities.GetChildItemType(listType);
  var isExtendedBindingList = typeof(IExtendedBindingList).IsAssignableFrom(listType);
  return (listType, childType, isExtendedBindingList);
}
```

Reply:
Fixed

Reply:
Nearly. Please remove the now obsolete Variables and the outcommented code.

Reply:
Done.

Reply:
Where are the variables from line 304 to 306 used? They should have unused and hence removed.

Reply:
Done